test/extensions/tracers/xray/localized_sampling_test.cc

```diff
@@ -0,0 +1,532 @@
+using testing::Return;
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace XRay {
+
+namespace {
+
+class LocalizedSamplingStrategyTest : public ::testing::Test {
+protected:
+  Event::SimulatedTimeSystem time_system_;
+};
+
+TEST_F(LocalizedSamplingStrategyTest, EmptyRules) {
+  NiceMock<Runtime::MockRandomGenerator> random_generator;
+  LocalizedSamplingStrategy strategy{"", random_generator, time_system_};
+  ASSERT_TRUE(strategy.usingDefaultManifest());
+}
+
+TEST_F(LocalizedSamplingStrategyTest, ValidCustomRules) {
+  NiceMock<Runtime::MockRandomGenerator> random_generator;
+  constexpr auto rules_json = R"EOF(
+{
+  "version": 2,
+  "rules": [
+    {
+      "description": "X-Ray rule",
+      "host": "*",
+      "http_method": "*",
+      "url_path": "/api/move/*",
+      "fixed_target": 0,
+      "rate": 0.05
+    }
+  ],
+  "default": {
+    "fixed_target": 1,
+    "rate": 0.1
+  }
+}
+  )EOF";
+  LocalizedSamplingStrategy strategy{rules_json, random_generator, time_system_};
+  ASSERT_FALSE(strategy.usingDefaultManifest());
+}
+
+TEST_F(LocalizedSamplingStrategyTest, DefaultRuleMissingRate) {
+  NiceMock<Runtime::MockRandomGenerator> random_generator;
+  constexpr auto rules_json = R"EOF(
+{
+  "version": 2,
+  "rules": [
+    {
+      "description": "X-Ray rule",
+      "host": "*",
+      "http_method": "*",
+      "url_path": "/api/move/*",
+      "fixed_target": 0,
+      "rate": 0.05
+    }
+  ],
+  "default": {
+    "fixed_target": 1
+  }
+}
+  )EOF";
+  LocalizedSamplingStrategy strategy{rules_json, random_generator, time_system_};
+  ASSERT_TRUE(strategy.usingDefaultManifest());
+}
+
+TEST_F(LocalizedSamplingStrategyTest, DefaultRuleMissingFixedTarget) {
+  NiceMock<Runtime::MockRandomGenerator> random_generator;
+  constexpr auto rules_json = R"EOF(
+{
+  "version": 2,
+  "rules": [
+    {
+      "description": "X-Ray rule",
+      "host": "*",
+      "http_method": "*",
+      "url_path": "/api/move/*",
+      "fixed_target": 0,
+      "rate": 0.05
+    }
+  ],
+  "default": {
+    "rate": 0.5
+  }
+}
+  )EOF";
+  LocalizedSamplingStrategy strategy{rules_json, random_generator, time_system_};
+  ASSERT_TRUE(strategy.usingDefaultManifest());
+}
+
+TEST_F(LocalizedSamplingStrategyTest, WrongVersion) {
+  NiceMock<Runtime::MockRandomGenerator> random_generator;
+  constexpr auto wrong_version = R"EOF(
+{
+  "version": 1,
+  "rules": [
+    {
+      "description": "X-Ray rule",
+      "host": "*",
+      "http_method": "*",
+      "url_path": "/api/move/*",
+      "fixed_target": 0,
+      "rate": 0.05
+    }
+  ],
+  "default": {
+    "fixed_target": 1,
+    "rate": 0.5
+  }
+}
+  )EOF";
+  LocalizedSamplingStrategy strategy{wrong_version, random_generator, time_system_};
+  ASSERT_TRUE(strategy.usingDefaultManifest());
+}
+
+TEST_F(LocalizedSamplingStrategyTest, MissingVersion) {
+  NiceMock<Runtime::MockRandomGenerator> random_generator;
+  constexpr auto missing_version = R"EOF(
+{
+  "version": 1,
```

Comment:
Is `"version"` supposed to be omitted from the config in this test?

Reply:
Good catch. Thank you.